### PR TITLE
PBKDF2 into FIPS boundary

### PR DESCRIPTION
**Pull request by Taffer**
### Issues:
Resolves CryptoAlg-1219

### Description of changes: 
* Moves PBKDF2 into the FIPS boundary
* Adds an RFC 6070 test that has sufficient password/salt for FIPS

### Call-outs:
N/A

### Testing:
Tested via `run_tests` on x86/aarch64.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: crypto/pkcs8/pkcs8_test.cc [hunk 1]
```diff
@@ -268,10 +268,13 @@ static void TestRoundTrip(int pbe_nid, const EVP_CIPHER *cipher,
 }
 
 TEST(PKCS8Test, DecryptString) {
-  TestDecrypt(kEncryptedPBES2WithDESAndSHA1,
-              sizeof(kEncryptedPBES2WithDESAndSHA1), "testing");
-  TestDecrypt(kEncryptedPBES2WithAESAndSHA256,
-              sizeof(kEncryptedPBES2WithAESAndSHA256), "testing");
+  if (!FIPS_mode()) {
```
Comment (nebeid) [inline crypto/pkcs8/pkcs8_test.cc]: Same comment as on pkcs12_test.cc

File: crypto/fipsmodule/pbkdf/pbkdf.c [hunk 2]
```diff
@@ -72,6 +73,21 @@ int PKCS5_PBKDF2_HMAC(const char *password, size_t password_len,
...
   HMAC_CTX_init(&hctx);
 
+  // We have to avoid the underlying SHA services updating the indicator
+  // state, so we lock the state here.
+  FIPS_service_indicator_lock_state();
+
+  // FIPS 140 parameter requirements, per NIST SP800-132:
+  //
+  // * key_len >= 14 bytes (112 bits)
+  // * salt_len >= 16 bytes (128 bits), assuming its randomly generated
```
[the remark this answers is not in the document]
Comment (Taffer) [reply crypto/fipsmodule/pbkdf/pbkdf.c:83]: Oh my, that's embarrassing.

File: crypto/pkcs8/pkcs8_test.cc (hunk 1 above)
Comment (Taffer) [reply crypto/pkcs8/pkcs8_test.cc]: CryptoAlg-1270

File: crypto/pkcs8/pkcs12_test.cc [hunk 3]
```diff
@@ -115,20 +115,24 @@ TEST(PKCS12Test, TestWindows) {
...
-  // PRF. (There is currently no way to specify the PRF in the pkcs12 command.)
-  std::string data = GetTestData("crypto/pkcs8/test/pbes2_sha1.p12");
-  TestImpl("kPBES2WithSHA1", StringToBytes(data), kPassword, nullptr);
-
-  // pbes2_sha256.p12 is a PKCS#12 file using PBES2 and HMAC-SHA-256. It was
-  // generated in the same way as pbes2_sha1.p12, but using OpenSSL 1.1.1b,
-  // which uses hmacWithSHA256 as the PRF.
-  data = GetTestData("crypto/pkcs8/test/pbes2_sha256.p12");
-  TestImpl("kPBES2WithSHA256", StringToBytes(data), kPassword, nullptr);
+  if (!FIPS_mode()) {
```
[the remark this answers is not in the document]
Comment (Taffer) [reply crypto/pkcs8/pkcs12_test.cc]: I've backed this out in favour of making the service indicator do something appropriate when you use invalid parameters (part of CryptoAlg-1270).

File: crypto/fipsmodule/pbkdf/pbkdf.c [hunk 4]
```diff
@@ -72,6 +73,18 @@ int PKCS5_PBKDF2_HMAC(const char *password, size_t password_len,
   HMAC_CTX hctx;
   HMAC_CTX_init(&hctx);
 
+  // We have to avoid the underlying SHA services updating the indicator
+  // state, so we lock the state here.
+  FIPS_service_indicator_lock_state();
+
+  // FIPS 140 parameter requirements, per NIST SP800-132:
```
Comment (nebeid) [inline crypto/fipsmodule/pbkdf/pbkdf.c:80]: Do we need to keep this comment here or it will go in the service indicator check?

**Review by nebeid — COMMENTED** [review]
I think the PR description needs editing since some work is moved to another upcoming PR.

File: crypto/fipsmodule/pbkdf/pbkdf.c (hunk 4 above)
Comment (Taffer) [reply crypto/fipsmodule/pbkdf/pbkdf.c:80]: Oh, good point. I wanted to keep it in the code somewhere, I'll see if there's a spot in the service indicator that works.

**Review by nebeid — APPROVED** [review]

**Review by DominicDams — APPROVED** [review]

**Pull request merged**